### PR TITLE
Add Chef Infra Client 17 release notes

### PR DESCRIPTION
**Pull request by tas50**
Signed-off-by: Tim Smith <tsmith@chef.io>

### CODE REVIEW

File: RELEASE_NOTES.md [hunk 1]
```diff
@@ -1,42 +1,263 @@
...
+
+Chef Infra Client can be configured to run as a scheduled task using the [chef-client cookbook](https://supermarket.chef.io/cookbooks/chef-client) or ideally using the [chef_client_scheduled_task resource](https://docs.chef.io/resources/chef_client_scheduled_task/) built into Chef Infra Client 16 or later. For users already running as a service setting up the scheduled task and then stopping the existing service can be performed within a Chef Infra Client run to migrate systems.
+
+#### Gem Resource Ruby 1.9+
+
+The `gem` resource used to install Ruby Gems into the system's Ruby installation will now assume Ruby 1.9 or later. As Ruby 1.8 or below have been end of life for nearly 7 years we believe there is little to no impact in this change.
+
+#### Legacy node['filesystem2'] removed on AIX/Solaris/FreeBSD
+
+The legacy `node['filesystem2']` attributes leftover from our multi-year migration of filesystem data on AIX, Solaris, and FreeBSD systems has been removed. This same data is now available at `node['filesystem']`
```
Comment (jaymzh) [inline RELEASE_NOTES.md:49]: what are we doing for windows? We probably want to go ahead and write new data to both places before we release 17 and make that the breaking change so in 18 we can drop fs2

File: RELEASE_NOTES.md [hunk 2]
```diff
@@ -1,42 +1,263 @@
...
+
+#### systemd_unit
+
+The `systemd_unit` resource has been improved to only shell out a single time to determine the state of the systemd unit. This optimization can result in significant performance improvements when using large numbers of `systemd_unit` resources. Thanks [@joshuamiller01](https://github.com/joshuamiller01)!
+
+#### execute
+
+The `execute` resource includes a new `login` property allowing you to run commands with a login shell. This is useful to ensure you have all potential environment variables defined in the user's shell.
 
 ### Ohai
```
Comment (jaymzh) [inline RELEASE_NOTES.md:119]: the cpu plugin was nearly completley rewritten by @ramereth  and includes cpu vuln info and supports a bunch of other new data that I'm sure he can tell you about.

File: RELEASE_NOTES.md [hunk 3]
```diff
@@ -1,42 +1,263 @@
...
+
+#### chef_client_launchd / macosx_service
+
+The `chef_client_launchd` and `macosx_service` resources have been updated to use the full path to the `launchctl` command. This avoids failures running these resources with incorrect PATH environment variables. Thanks for this improvement [@krackajak](https://github.com/krackajak)!
+
+#### systemd_unit
+
+The `systemd_unit` resource has been improved to only shell out a single time to determine the state of the systemd unit. This optimization can result in significant performance improvements when using large numbers of `systemd_unit` resources. Thanks [@joshuamiller01](https://github.com/joshuamiller01)!
+
+#### execute
```
Comment (jaymzh) [inline RELEASE_NOTES.md:115]: I think some of these landed in late 16 versions? If so, do we want to call out some of the windows_user fixes that landed in late 16 versions? No strong opinion here, just thought it was worth mentioning.

File: RELEASE_NOTES.md (hunk 2 above)
Comment (tas50) [reply RELEASE_NOTES.md:119]: Added

File: RELEASE_NOTES.md (hunk 1 above)
Comment (tas50) [reply RELEASE_NOTES.md:49]: That's not a thing that's in 17

File: RELEASE_NOTES.md (hunk 3 above)
Comment (tas50) [reply RELEASE_NOTES.md:115]: The file verify optimizations went in, but I don't believe this was backported

File: RELEASE_NOTES.md (hunk 1 above)
Comment (jaymzh) [reply RELEASE_NOTES.md:49]: but it COULD be. And it should be, so we're not still dealing with this in 19

File: RELEASE_NOTES.md [hunk 4]
```diff
@@ -1,42 +1,271 @@
...
+
+#### 32bit RHEL/CentOS 6 Support
+
+Chef Infra Client 17 packages will no longer be produced for 32bit RHEL/CentOS 6 systems. RHEL/CentOS 6 became EOL November 2020. We are extending support for 64 bit RHEL/CentOS 6 until Chef Infra Client 18 (April 2022) or when upstream platform or library changes prohibit us from building on these EOL systems.
+
+#### Chef Client As A Service on Windows
+
+Based on customer feedback and observations in the field we've removed the ability to run the Chef Infra Client as a service on Windows nodes. We've seen the service manager for the Chef Infra Client consume excessive memory, hang preventing runs, or prevent nodes from updating to new client releases properly. We've always seen significant better reliability by running Chef Infra Client as a scheduled task on Windows and in July of 2006 we introduced warnings to the [chef-client cookbook](https://supermarket.chef.io/cookbooks/chef-client) when running as a service. The ability to setup the client as a service was later removed from the cookbook entirely in October of 2017.
```
Comment (jasonwbarnett) [inline RELEASE_NOTES.md:37]:
```suggestion
Based on customer feedback and observations in the field we've removed the ability to run the Chef Infra Client as a service on Windows nodes. We've seen the service manager for the Chef Infra Client consume excessive memory, hang preventing runs, or prevent nodes from updating to new client releases properly. We've always seen significantly better reliability by running Chef Infra Client as a scheduled task on Windows and in July of 2006 we introduced warnings to the [chef-client cookbook](https://supermarket.chef.io/cookbooks/chef-client) when running as a service. The ability to setup the client as a service was later removed from the cookbook entirely in October of 2017.
```

File: RELEASE_NOTES.md (hunk 1 above)
Comment (jaymzh) [reply RELEASE_NOTES.md:49]: At the very least, we should solidify a timeline, here and now.
Note that `filesystem2` was added to Windows, that both fs and fs2 will be the same in Chef 18, and that fs2 will be dropped from Windows in Chef 19.
Users need to know what to expect.

Comment (NaomiReeves) [reply RELEASE_NOTES.md:49]: I'll second this. Prolonging this conversion has been tough for our (Facebook) cross platform cookbooks, especially with only one remaining platform continuing to have a default of the old-style data. If we can't start this conversion with 17, we definitely need a committed timeline so people can plan for it.

File: RELEASE_NOTES.md [hunk 5]
```diff
@@ -1,42 +1,279 @@
...
+
+For customers current running Chef Infra Client as a service we suggest migrating to scheduled task based execution. This allows for complex scheduling scenarios not possible with simple services such, as the ability to skip Chef Infra Client execution if a system is on battery or to run the Chef Infra Client as soon as a system boots to ensure configuration.
+
+Chef Infra Client can be configured to run as a scheduled task using the [chef-client cookbook](https://supermarket.chef.io/cookbooks/chef-client) or ideally using the [chef_client_scheduled_task resource](https://docs.chef.io/resources/chef_client_scheduled_task/) built into Chef Infra Client 16 or later. For users already running as a service setting up the scheduled task and then stopping the existing service can be performed within a Chef Infra Client run to migrate systems.
```
Comment (kagarmoe) [inline RELEASE_NOTES.md:41]: If using the chef_client_scheduled_task resource is the preferred solution, then we should list it first.

File: RELEASE_NOTES.md [hunk 6]
```diff
@@ -1,42 +1,279 @@
...
+
+The `gem` resource used to install Ruby Gems into the system's Ruby installation will now assume Ruby 1.9 or later. As Ruby 1.8 or below have been end of life for nearly 7 years we believe there is little to no impact in this change.
+
+#### Legacy node['filesystem2'] removed on AIX/Solaris/FreeBSD
+
+The legacy `node['filesystem2']` attributes leftover from our multi-year migration of filesystem data on AIX, Solaris, and FreeBSD systems has been removed. This same data is now available at `node['filesystem']`
+
+#### Removed Antergos and Pidora Detection
+
+Ohai detection of the end of life Antergos and Pidora distributions has been removed. Antergos ended releases and downloads of the distribution in May 2019 and Pidora stops receiving updates in 2014.
```
Comment (kagarmoe) [inline RELEASE_NOTES.md:53]:
```suggestion
Ohai detection of the end-of-life Antergos and Pidora distributions has been removed. Antergos ended releases and downloads of the distribution in May 2019 and Pidora stopped receiving updates in 2014.
```

File: RELEASE_NOTES.md [hunk 7]
```diff
@@ -1,42 +1,279 @@
...
+A common problem when using the "wrapper cookbook" pattern is when the wrapped cookbook declares what are called "derived attributes", which are attributes that refer to other attributes. Because of the order that attribute files are parsed in, this does not work as intended when the base attribute is changed in a wrapper cookbook. By extending the use of the `lazy {}` helper to the declaration of node attributes, it makes it possible for the wrapped cookbook to cleanly allow wrapper cookbooks to override base attributes as intended.
+
+Instead of:
+
+```ruby
+default['myapp']['dir'] = '/opt/myapp'
+default['myapp']['bindir'] = "#{node['myapp']['dir']}/bin"
+```
+
+Use the lazy helper:
```
Comment (kagarmoe) [inline RELEASE_NOTES.md:68]: Move the lazy helper example above the less-desired pattern. Show people what you want them to do before showing them what you don't want them to do.

File: RELEASE_NOTES.md [hunk 8]
```diff
@@ -1,42 +1,279 @@
...
+default['myapp']['bindir'] = lazy { "#{node['myapp']['dir']}/bin" }
+```
+
+The wrapper cookbook can then override the base attribute and the derived attribute will change:
+
+```ruby
+default['myapp']['dir'] = "/opt/my_better_app" # this also changes the bindir attribute correctly
+```
+
+The use of this helper is not limited to declarations in attribute files and can be used whenever attributes are being assigned, but it is better left to this pattern.
```
Comment (kagarmoe) [inline RELEASE_NOTES.md:81]: I'm confused. Should I only use it in wrappers or not?

File: RELEASE_NOTES.md [hunk 9]
```diff
@@ -1,42 +1,279 @@
...
+
+The `apt_package` resource now properly handles downgrading package versions. Please note that full versions must be provided in the `version` property and invalid version strings will now raise and error. Thanks for this improvement [@jaymzh](https://github.com/jaymzh)!
+
+#### chef_client_launchd / macosx_service
+
+The `chef_client_launchd` and `macosx_service` resources have been updated to use the full path to the `launchctl` command. This avoids failures running these resources with incorrect PATH environment variables. Thanks for this improvement [@krackajak](https://github.com/krackajak)!
+
+#### execute
+
+The `execute` resource includes a new `login` property allowing you to run commands with a login shell. This is useful to ensure you have all potential environment variables defined in the user's shell.
```
Comment (kagarmoe) [inline RELEASE_NOTES.md:113]:
```suggestion
The `execute` resource includes a new `login` property allowing you to run commands with a login shell. This helps ensure you have all potential environment variables defined in the user's shell.
```

File: RELEASE_NOTES.md [hunk 10]
```diff
@@ -1,42 +1,279 @@
...
+
+#### hostname
+
+The `hostname` resource now includes a new `fqdn` property to allow you to set a custom fqdn in the hostname file in addition to the system's hostname. Thanks for suggesting this improvement [@evandam](https://github.com/evandam)!
+
+#### systemd_unit
+
+The `systemd_unit` resource has been improved to only shell out a single time to determine the state of the systemd unit. This optimization can result in significant performance improvements when using large numbers of `systemd_unit` resources. Thanks [@joshuamiller01](https://github.com/joshuamiller01)!
```
Comment (kagarmoe) [inline RELEASE_NOTES.md:121]:
```suggestion
The `systemd_unit` resource has been improved to only shell out once to determine the state of the systemd unit. This optimization should result in significant performance improvements when using large numbers of `systemd_unit` resources. Thanks [@joshuamiller01](https://github.com/joshuamiller01)!
```

File: RELEASE_NOTES.md [hunk 11]
```diff
@@ -1,42 +1,279 @@
...
+
+#### windows_certificate
+
+The `windows_certificate` resource has seen a large overhaul with improved support for importing and exporting certificate objects, the ability to create certificates objects from a URL and a new `output_path` property used when exporting.
```
Comment (kagarmoe) [inline RELEASE_NOTES.md:125]:
```suggestion
The `windows_certificate` resource has undergone a large overhaul, with improved support for importing and exporting certificate objects, the ability to create certificate objects from a URL, and a new `output_path` property for use with exporting.
```

File: RELEASE_NOTES.md [hunk 12]
```diff
@@ -1,42 +1,279 @@
...
+
+#### windows_task
+
+The `windows_task` resource now has a new `backup` property that allows you to control the number of XML backups that will be kept of your Windows Scheduled Task definition when modified. This default to `5` and can be disabled by setting the property to `false`. Thanks [@ kimbernator](https://github.com/kimbernator)!
```
Comment (kagarmoe) [inline RELEASE_NOTES.md:129]:
```suggestion
The `windows_task` resource now has a new `backup` property that allows you to control the number of XML backups that will be kept of your Windows Scheduled Task definition. This default for this setting is `5` and can be disabled by setting the property to `false`. Thanks [@ kimbernator](https://github.com/kimbernator)!
```

Comment (kagarmoe) [reply RELEASE_NOTES.md:129]: I wish I had thought of `kimbernator` as my handle.

File: RELEASE_NOTES.md [hunk 13]
```diff
@@ -1,42 +1,279 @@
...
+  "system": "podman",
+  "role": "guest"
+}
+```
+
+Thanks for this addition [@ramereth](https://github.com/ramereth)!
+
+#### Habitat Support
+
+Ohai includes a new `:Habitat` plugin that gathers information on the Habitat installation including the version of Habitat installed, installed packages, and running services.
```
Comment (kagarmoe) [inline RELEASE_NOTES.md:165]:
```suggestion
Ohai includes a new `:Habitat` plugin that gathers information about the Habitat installation, including installed Habitat version, installed packages, and running services.
```

File: RELEASE_NOTES.md [hunk 14]
```diff
@@ -1,42 +1,279 @@
...
+
+For customers current running Chef Infra Client as a service we suggest migrating to scheduled task based execution. This allows for complex scheduling scenarios not possible with simple services such, as the ability to skip Chef Infra Client execution if a system is on battery or to run the Chef Infra Client as soon as a system boots to ensure configuration.
+
+Chef Infra Client can be configured to run as a scheduled task using the [chef-client cookbook](https://supermarket.chef.io/cookbooks/chef-client) or ideally using the [chef_client_scheduled_task resource](https://docs.chef.io/resources/chef_client_scheduled_task/) built into Chef Infra Client 16 or later. For users already running as a service setting up the scheduled task and then stopping the existing service can be performed within a Chef Infra Client run to migrate systems.
+
+#### Gem Resource Ruby 1.9+
+
+The `gem` resource used to install Ruby Gems into the system's Ruby installation will now assume Ruby 1.9 or later. As Ruby 1.8 or below have been end of life for nearly 7 years we believe there is little to no impact in this change.
```
Comment (kagarmoe) [inline RELEASE_NOTES.md:45]:
```suggestion
The `gem` resource used to install Ruby Gems into the system's Ruby installation will now assume Ruby 1.9 or later. As Ruby 1.8 and below reached end of life almost 7 years ago, we believe there is little to no impact in this change.
```

**Review by kagarmoe — CHANGES_REQUESTED** [review]
These are quite good. I have tons of nits, of course.

File: RELEASE_NOTES.md [hunk 15]
```diff
@@ -1,42 +1,279 @@
...
+
+#### Legacy node['filesystem2'] removed on AIX/Solaris/FreeBSD
+
+The legacy `node['filesystem2']` attributes leftover from our multi-year migration of filesystem data on AIX, Solaris, and FreeBSD systems has been removed. This same data is now available at `node['filesystem']`
+
```
Comment (jaymzh) [inline RELEASE_NOTES.md:50]: Given that https://github.com/chef/ohai/pull/1662 is now merged, we should probably amend this like so...

```suggestion
#### Legacy node['filesystem2'] removed on AIX/Solaris/FreeBSD

The legacy `node['filesystem2']` attributes leftover from our multi-year migration of filesystem data on AIX, Solaris, and FreeBSD systems has been removed. This same data is now available at `node['filesystem']`

#### node['filesystem'] on Windows updated to new-style data

`node['filesystem']` on Windows now has what `node['filesytem2']` has had on Windows throughout Chef 16. Both node attributes now have the same data as a way to allow users to transition from `filesystem2` to `filesystem` safely. In Chef 18, we will remove `node['filesystem2']` completely thus finishing our multi-year migration of Ohai filesystem data format.
```

File: RELEASE_NOTES.md [hunk 16]
```diff
@@ -1,42 +1,279 @@
...
+### Knife Moved to Workstation
+
+For historical packaging reasons the Chef Infra Client packages have always shipped with the `knife` command for managing your Chef Infra nodes. With Chef Workstation there's no benefit to shipping knife in the Chef Infra Client package and there are several downsides. Shipping management tooling within the client is seen as a security risk to many and increases the side of the Chef Infra Client codebase by adding a large number of management dependencies. With Chef Infra Client 17 we've split knife into it own Ruby Gem, which will continue to ship in Chef Workstation, but will no longer come bundled with Chef Infra Client. We hope you'll enjoy the new faster and smaller Chef Infra Client while continuing to use knife in Chef Workstation uninterrupted.
+
+### Breaking Changes
+
+#### AIX Virtualization Improvements
+
+The Ohai :Virtualization plugin on AIX systems will now properly return the `lpar_no` and `wpar_no` values as Integers instead of Strings. This makes the data much easier to work with in cookbooks, but may be a breaking change depending on how AIX users consumed these values.
```
[the remark this answers is not in the document]
Comment (tas50) [reply RELEASE_NOTES.md:29]: Yes. That's how Ohai plugins are named

File: RELEASE_NOTES.md (hunk 1 above)
Comment (tas50) [reply RELEASE_NOTES.md:49]: That fix made it in. I'll update the docs

File: RELEASE_NOTES.md [hunk 17]
```diff
@@ -1,42 +1,279 @@
...
-- Improved performance in systemd_unit resource - https://github.com/chef/chef/pull/10925
-- gem resource: assume rubygems 1.8+ now: https://github.com/chef/chef/pull/10379
-- execute: Add login property - https://github.com/chef/chef/pull/11201
+#### Logging Improvements
+
+A large number of resources have seen improvements to the logging available in the `debug` log level providing better information for troubleshooting Chef Infra Client execution. Thanks for this improvement [@jaymzh](https://github.com/jaymzh)!
+
+#### apt_package
+
+The `apt_package` resource now properly handles downgrading package versions. Please note that full versions must be provided in the `version` property and invalid version strings will now raise and error. Thanks for this improvement [@jaymzh](https://github.com/jaymzh)!
```
[the remark this answers is not in the document]
Comment (tas50) [reply RELEASE_NOTES.md]: Fixed. Thanks

File: RELEASE_NOTES.md (hunk 5 above)
Comment (tas50) [reply RELEASE_NOTES.md:41]: The reason for mentioning the chef-client cookbook is just to show that this is not a new warning.

File: RELEASE_NOTES.md (hunk 7 above)
Comment (tas50) [reply RELEASE_NOTES.md:71]: Done

File: RELEASE_NOTES.md (hunk 8 above)
Comment (tas50) [reply RELEASE_NOTES.md:84]: Nuking some of that warning

**Pull request merged**